pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go

```diff
@@ -150,9 +153,91 @@ func TestFindAndAddNewPods_WithRescontructedVolume(t *testing.T) {
 
 }
 
+func TestFindAndAddNewPods_WithDifferentConditions(t *testing.T) {
+	tests := []struct {
+		desc          string
+		hasAddedPods  bool
+		podState      string
+		expectedFound bool // Found pod is added to DSW
+	}{
+		{
+			"HasAddedPods is false, ShouldPodRuntimeBeRemoved is true",
+			false,
+			"removed",
+			false, // Pod should not be added to DSW
+		},
+		{
+			"HasAddedPods is false, ShouldPodRuntimeBeRemoved is false",
+			false,
+			"other",
+			true, // Pod should be added to DSW
+		},
+		{
+			"HasAddedPods is true, ShouldPodContainersBeTerminating is true",
+			true,
+			"terminating",
+			false, // Pod should not be added to DSW
+		},
+		{
+			"HasAddedPods is true, ShouldPodContainersBeTerminating is false",
+			true,
+			"other",
+			true, // Pod should be added to DSW
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.desc, func(t *testing.T) {
+			// create dswp
+			dswp, fakePodManager, fakePodState := prepareDswpWithVolume(t)
+
+			// create pod
+			containers := []v1.Container{
+				{
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      "dswp-test-volume-name",
+							MountPath: "/mnt",
+						},
+					},
+				},
+			}
+			pod := createPodWithVolume("dswp-test-pod", "dswp-test-volume-name", "file-bound", containers)
+
+			fakePodManager.AddPod(pod)
+
+			switch tc.podState {
+			case "removed":
```

Comment:
I think in the real podStateManager, "removed" is also counted as "terminating"

Reply:
They check differently though?
One is checking whether the podId is in the list of PodState.removed.
The other one is checking whether the podId is in the list of PodState.terminating.

Reply:
In the real implementation, it checks for `IsTerminationStarted`
https://github.com/kubernetes/kubernetes/blob/126986016e363c5dd92a202e3f433f6696dbcae1/pkg/kubelet/pod_workers.go#L491

I believe this returns true even when the pod has finished terminating. @kubernetes/sig-node-pr-reviews for confirmation.